akka-stream/src/main/scala/akka/stream/scaladsl/SourceRef.scala

```diff
@@ -0,0 +1,272 @@
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case SubscriptionTimeoutTimerKey ⇒
+          val ex = StreamRefs.StreamRefSubscriptionTimeoutException(
+            // we know the future has been competed by now, since it is in preStart
+            s"[$stageActorName] Remote side did not subscribe (materialize) handed out Sink reference [${promise.future.value}]," +
+              s"within subscription timeout: ${PrettyDuration.format(subscriptionTimeout.timeout)}!")
+
+          throw ex // this will also log the exception, unlike failStage; this should fail rarely, but would be good to have it "loud"
+
+        case DemandRedeliveryTimerKey ⇒
+          log.debug("[{}] Scheduled re-delivery of demand until [{}]", stageActorName, localCumulativeDemand)
+          getPartnerRef ! StreamRefs.CumulativeDemand(localCumulativeDemand)
+          scheduleDemandRedelivery()
+      }
+
+      lazy val initialReceive: ((ActorRef, Any)) ⇒ Unit = {
+        case (sender, msg @ StreamRefs.OnSubscribeHandshake(remoteRef)) ⇒
+          cancelTimer("SubscriptionTimeoutTimerKey")
+          observeAndValidateSender(remoteRef, "Illegal sender in SequencedOnNext")
+          log.debug("[{}] Received handshake {} from {}", stageActorName, msg, sender)
+
+          triggerCumulativeDemand()
+
+        case (sender, msg @ StreamRefs.SequencedOnNext(seqNr, payload)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in SequencedOnNext")
+          observeAndValidateSequenceNr(seqNr, "Illegal sequence nr in SequencedOnNext")
+          log.debug("[{}] Received seq {} from {}", stageActorName, msg, sender)
+
+          tryPush(payload)
+          triggerCumulativeDemand()
+
+        case (sender, StreamRefs.RemoteStreamCompleted(seqNr)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in RemoteSinkCompleted")
+          observeAndValidateSequenceNr(seqNr, "Illegal sequence nr in RemoteSinkCompleted")
+          log.debug("[{}] The remote stream has completed, completing as well...", stageActorName)
+
+          self.unwatch(sender)
+          completeStage()
+
+        case (sender, StreamRefs.RemoteStreamFailure(reason)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in RemoteSinkFailure")
+          log.warning("[{}] The remote stream has failed, failing (reason: {})", stageActorName, reason)
+
+          self.unwatch(sender)
+          failStage(StreamRefs.RemoteStreamRefActorTerminatedException(s"Remote stream (${sender.path}) failed, reason: $reason"))
+
+        case (_, Terminated(ref)) ⇒
+          if (getPartnerRef == ref)
+            failStage(StreamRefs.RemoteStreamRefActorTerminatedException(s"The remote partner ${getPartnerRef} has terminated! " +
+              s"Tearing down this side of the stream as well."))
+        // else this should not have happened, and we ignore this -- someone may have been doing weird things!
+      }
+
+      def tryPush(): Unit =
```

Comment:
A simpler version of both `tryPush` methods:

```scala
      def tryPush(): Unit =
        if (receiveBuffer.nonEmpty)
          push(out, receiveBuffer.dequeue())

      private def onReceiveElement(payload: T): Unit = {
        localRemainingRequested -= 1
        if (receiveBuffer.isEmpty && isAvailable(out))
          push(out, payload)
        else if (receiveBuffer.isFull)
          throw new IllegalStateException(s"Attempted to overflow buffer! Capacity: ${receiveBuffer.capacity}, incoming element: $payload, localRemainingRequested: ${localRemainingRequested}, localCumulativeDemand: ${localCumulativeDemand}")
        else
          receiveBuffer.enqueue(payload)
      }
```

Reply:
Got it, thanks!